src/ripple/rpc/handlers/GatewayBalances.cpp

```diff
@@ -0,0 +1,228 @@
+
+    // Parse the specified hotwallet(s), if any
+    std::set <Account> hotWallets;
+
+    if (params.isMember ("hotwallet"))
+    {
+        Json::Value const& hw = params["hotwallet"];
+        bool valid = true;
+
+        auto addHotWallet = [&valid, &hotWallets](Json::Value const& j)
+        {
+            if (j.isString())
+            {
+                RippleAddress ra;
+                if (! ra.setAccountPublic (j.asString ()) &&
+                    ! ra.setAccountID (j.asString()))
+                {
+                    valid = false;
+                }
+                else
+                    hotWallets.insert (ra.getAccountID ());
+            }
+            else
+            {
+                valid = false;
+            }
+        };
+
+        if (hw.isArray())
+        {
+            for (unsigned i = 0; i < hw.size(); ++i)
+                addHotWallet (hw[i]);
+        }
+        else if (hw.isString())
+        {
+            addHotWallet (hw);
+        }
+        else
+        {
+            valid = false;
+        }
+
+        if (! valid)
+        {
+            result[jss::error]   = "invalidHotWallet";
+            return result;
+        }
+
+    }
+
+    std::map <Currency, STAmount> sums;
+    std::map <Account, std::vector <STAmount>> hotBalances;
+    std::map <Account, std::vector <STAmount>> assets;
+
+    // Traverse the cold wallet's trust lines
+    ledger->visitAccountItems (accountID, [&](SLE::ref sle)
+    {
+        if (sle->getType() == ltRIPPLE_STATE)
+        {
+            RippleState rs (sle, accountID);
+
+            int balSign = rs.getBalance().signum();
+            if (balSign == 0)
+                return;
+
+            auto const& peer = rs.getAccountIDPeer();
+
+            // Here, a negative balance means the cold wallet owes (normal)
+            // A positive balance means the cold wallet has an asset (unusual)
+
+            if (hotWallets.count (peer) > 0)
+            {
+                // This is a specified hot wallt
+                hotBalances[peer].push_back (-rs.getBalance ());
+            }
+            else if (balSign > 0)
+            {
+                // This is a gateway asset
+                assets[peer].push_back (rs.getBalance ());
+            }
+            else
+            {
+                // normal negative balance, obligation to customer
+                auto& bal = sums[rs.getBalance().getCurrency()];
+                if (bal == zero)
+                {
+                    // This is needed to set the currency code correctly
+                    bal = -rs.getBalance();
+                }
+                else
+                    bal -= rs.getBalance();
+            }
+        }
+    });
+
+    if (! sums.empty())
+    {
+        Json::Value& j = (result [jss::obligations] = Json::objectValue);
+        for (auto const& e : sums)
+        {
+            j[to_string (e.first)] = e.second.getText ();
+        }
+    }
+
+    if (! hotBalances.empty())
```

Comment:
These two code blocks coming up seem exactly the same.  Couldn't they be made into one?